src/Path/Module.hs

```diff
@@ -1,95 +1,148 @@
+  , declTerm :: Spanned a
+  , declType :: Spanned a
   }
   deriving (Eq, Foldable, Functor, Ord, Show, Traversable)
 
-newtype Import = Import { importModuleName :: ModuleName }
-  deriving (Eq, Ord, Show)
-
-data Decl v a
-  = Declare v a
-  | Define v a
-  | Doc String (Spanned (Decl v a))
-  deriving (Eq, Foldable, Functor, Ord, Show, Traversable)
-
-declName :: Decl v a -> v
-declName (Declare v _)    = v
-declName (Define  v _)    = v
-declName (Doc _ (d :~ _)) = declName d
-
-
-newtype ModuleGraph v a = ModuleGraph { unModuleGraph :: Map.Map ModuleName (Module v a) }
-  deriving (Eq, Ord, Show)
-
-moduleGraph :: [Module v a] -> ModuleGraph v a
-moduleGraph = ModuleGraph . Map.fromList . map ((,) . moduleName <*> id)
-
-modules :: ModuleGraph v a -> [Module v a]
-modules = Map.elems . unModuleGraph
-
-
-lookupModule :: (Carrier sig m, Member (Error Doc) sig) => ModuleGraph v a -> Spanned Import -> m (Module v a)
-lookupModule g (i :~ s) = maybe (unknownModule (i :~ s)) pure (Map.lookup (importModuleName i) (unModuleGraph g))
-
-cycleFrom :: (Carrier sig m, Effect sig, Member (Error Doc) sig) => ModuleGraph v a -> Spanned Import -> m ()
+newtype ModuleGraph f a = ModuleGraph { unModuleGraph :: Map.Map ModuleName (ScopeH Qualified (Module f) f a) }
+  deriving (Foldable, Functor, Monoid, Semigroup, Traversable)
+
+deriving instance (Eq   a, forall a . Eq   a => Eq   (f a), Monad f) => Eq   (ModuleGraph f a)
+deriving instance (Ord  a, forall a . Eq   a => Eq   (f a)
+                         , forall a . Ord  a => Ord  (f a), Monad f) => Ord  (ModuleGraph f a)
+deriving instance (Show a, forall a . Show a => Show (f a))          => Show (ModuleGraph f a)
+
+instance Monad f => RModule (ModuleGraph f) f where
+  ModuleGraph ms >>=* f = ModuleGraph (fmap (>>=* f) ms)
+
+moduleGraph :: Applicative f => [Module f Qualified] -> ModuleGraph f Void
+moduleGraph ms = ModuleGraph (Map.fromList (map ((,) . moduleName <*> bindHEither Left) ms))
+
+restrict :: Set.Set ModuleName -> ModuleGraph f a -> ModuleGraph f a
+restrict keys = ModuleGraph . flip Map.restrictKeys keys . unModuleGraph
+
+rename :: (Carrier sig m, Foldable t, Member (Error Doc) sig, Member (Reader Span) sig)
+       => t (Module f a)
+       -> User
+       -> m Qualified
+rename ms n = case foldMap (\ m -> [ moduleName m :.: n | d <- toList (moduleDecls m), declName d == n ]) ms of
+  [x]  -> pure x
+  []   -> freeVariables (pure n)
+  x:xs -> ambiguousName n (x:|xs)
+
+runDecl :: Carrier sig m => (a -> ReaderC Span m b) -> Decl a -> m (Decl b)
+runDecl f (Decl n d tm ty) = do
+  tm' <- runSpanned f tm
+  ty' <- runSpanned f ty
+  pure (Decl n d tm' ty')
+
+renameDecl :: (Carrier sig m, Foldable t, Member (Error Doc) sig, Traversable g)
+           => t (Module f a)
+           -> Decl (g User)
+           -> m (Decl (g Qualified))
+renameDecl ms = runDecl (traverse (rename ms))
+
+renameModule :: (Carrier sig m, Foldable t, Member (Error Doc) sig, Traversable g)
```

Comment:
Renaming modules, declarations, and individual names happens entirely in here (and via `Traversable`) now.